File: app/src/main/java/com/example/all_sensors/presentation/service/ForegroundService.java
package com.example.all_sensors.presentation.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.Build;
import android.os.IBinder;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import com.example.all_sensors.R;
import com.example.all_sensors.presentation.AccListener;
import com.example.all_sensors.presentation.HrListener;
import com.example.all_sensors.presentation.PpgListener;

import com.example.all_sensors.presentation.OnDemandScheduler;
import com.example.all_sensors.presentation.TrackerManager;
import com.example.all_sensors.presentation.data.DataBuffer;
import com.samsung.android.service.health.tracking.ConnectionListener;
import com.samsung.android.service.health.tracking.HealthTrackerException;
import com.samsung.android.service.health.tracking.data.HealthTrackerType;

import android.os.PowerManager;
import android.widget.Toast;

import java.io.IOException;

/** Service en primer plano que mantiene sensores continuos y lanza on-demand */
public class ForegroundService extends Service {
    private static final String CHANNEL_ID = "sensors_channel";
    private static final int NOTIF_ID = 1;
    private TrackerManager trackerManager;
    private OnDemandScheduler scheduler;
    private PowerManager.WakeLock wakeLock;
    private DataBuffer dataBuffer;

    @Override
    public void onCreate() {
        super.onCreate();

        dataBuffer = new DataBuffer();

        // PREPARA EL WAKELOCK (partial, solo CPU)
        PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK,"AllSensors:ForegroundWakelock");

        // 1) Creamos un ConnectionListener que arranca trackers al conectar
        ConnectionListener connListener = new ConnectionListener() {
            @Override
            public void onConnectionSuccess() {
                // Marca en el manager que ya estamos conectados
                trackerManager.markConnected();

                // ¡Ahora sí podemos arrancar los continuos!
                trackerManager.startContinuous(HealthTrackerType.ACCELEROMETER_CONTINUOUS,
                        new AccListener(dataBuffer));
                trackerManager.startContinuous(HealthTrackerType.HEART_RATE_CONTINUOUS,
                        new HrListener(dataBuffer));
                //trackerManager.startContinuous(HealthTrackerType.PPG_CONTINUOUS,
                        //new PpgListener(dataBuffer));

                // Y también arrancamos el scheduler on-demand
                scheduler.start();
            }
            @Override public void onConnectionFailed(HealthTrackerException e) {
                // Maneja fallo (quizá notificar con Toast o reintentar)
            }
            @Override public void onConnectionEnded() {
                trackerManager.markDisconnected();
            }


        };

        // 2) Instanciamos TrackerManager **con** ese listener
        trackerManager = new TrackerManager(this, connListener);

        // 3) Preparamos el scheduler, pero **no** lo arrancamos todavía
        scheduler = new OnDemandScheduler(trackerManager, dataBuffer);
    }


    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        // 4) Crea canal y notificación de primer plano
        createNotificationChannel();
        Notification notif = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("Sensores activos")
                .setContentText("Recolección continua y eventos on-demand")
                .setSmallIcon(R.drawable.ic_launcher_foreground)  // icono en mipmap o drawable
                .build();
        startForeground(NOTIF_ID, notif);

        // **5) Adquiere el WakeLock** para que el CPU siga despierto con pantalla apagada
        if (wakeLock != null && !wakeLock.isHeld()) {
            wakeLock.acquire();
        }

        // 6) Sticky para que se reinicie si Android mata el service
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        // **Libera el WakeLock**
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
        }

        // Para scheduler y trackers, desconecta SDK
        if (scheduler != null) scheduler.stop();
        if (trackerManager != null) {
            trackerManager.stopAllContinuous();
            trackerManager.shutdown();
        }
    }

    /** Binder para que la actividad pueda invocar exportCsv() */
    public class LocalBinder extends Binder {
        public ForegroundService getService() {
            return ForegroundService.this;
        }
    }
    private final IBinder binder = new LocalBinder();

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }

    /** Llamarás a esto desde tu UI para volcar a CSV */
    public void exportCsv(String filename) {
        try {
            dataBuffer.exportToCsv(this, filename);
            Toast.makeText(this, "CSV grabado: " + filename, Toast.LENGTH_SHORT).show();
        } catch (IOException e) {
            Toast.makeText(this, "Error exportando CSV", Toast.LENGTH_SHORT).show();
        }
    }


    /** Crea canal de notificación necesario en Android 8+ */
    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return;
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "Sensores en primer plano",
                NotificationManager.IMPORTANCE_LOW
        );
        channel.setDescription("Recolección continua de sensores y on-demand");
        NotificationManager nm = getSystemService(NotificationManager.class);
        if (nm != null) nm.createNotificationChannel(channel);
    }
}

File: app/src/main/java/com/signals/smartwatch/service/ForegroundService.kt
package com.signals.smartwatch.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import android.util.Log
import android.widget.Toast
import androidx.core.app.NotificationCompat
import com.signals.smartwatch.R
import com.signals.smartwatch.SensorEvents
import com.signals.smartwatch.data.CsvExporter
import com.signals.smartwatch.data.SensorDatabase
import com.signals.smartwatch.sensors.samsung.AccListener
import com.signals.smartwatch.sensors.samsung.HrListener
import com.signals.smartwatch.sensors.samsung.OnDemandScheduler
import com.signals.smartwatch.sensors.samsung.PpgListener
import com.signals.smartwatch.sensors.samsung.SkinTempListener
import com.signals.smartwatch.sensors.samsung.TrackerManager
import com.samsung.android.service.health.tracking.ConnectionListener
import com.samsung.android.service.health.tracking.HealthTrackerException
import com.samsung.android.service.health.tracking.data.HealthTrackerType
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException

class ForegroundService : Service() {

    // ── Coroutine scope del servicio ──────────────────────────────────────────
    private val serviceJob = SupervisorJob()
    private val serviceScope = CoroutineScope(Dispatchers.Main + serviceJob)

    // ── Sensores ──────────────────────────────────────────────────────────────
    private lateinit var trackerManager: TrackerManager
    private lateinit var scheduler: OnDemandScheduler

    // ── Backoff de reconexión ─────────────────────────────────────────────────
    @Volatile private var reconnectAttempt = 0

    // ── Room ──────────────────────────────────────────────────────────────────
    private val dao by lazy { SensorDatabase.getInstance(this).sensorDao() }

    // ── WakeLock ──────────────────────────────────────────────────────────────
    private var wakeLock: PowerManager.WakeLock? = null

    // ── Listener UI (vía LocalBinder) ─────────────────────────────────────────
    @Volatile private var uiListener: SensorEvents? = null

    private val listenerProxy = object : SensorEvents {
        override fun onAcc(x: Float, y: Float, z: Float) = uiListener?.onAcc(x, y, z) ?: Unit
        override fun onHr(bpm: Int)                       = uiListener?.onHr(bpm)       ?: Unit
        override fun onPpg(green: Float, ir: Float, red: Float) = uiListener?.onPpg(green, ir, red) ?: Unit
        override fun onSpO2(percent: Float)               = uiListener?.onSpO2(percent) ?: Unit
        override fun onSkinTemp(celsius: Float)           = uiListener?.onSkinTemp(celsius) ?: Unit
    }

    fun setUiListener(l: SensorEvents)  { uiListener = l }
    fun clearUiListener()               { uiListener = null }

    // ── Lifecycle ─────────────────────────────────────────────────────────────
    override fun onCreate() {
        super.onCreate()

        val pm = getSystemService(POWER_SERVICE) as PowerManager
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "AllSensors:ForegroundWakelock")

        val connListener = object : ConnectionListener {
            override fun onConnectionSuccess() {
                reconnectAttempt = 0
                trackerManager.markConnected()
                trackerManager.startContinuous(
                    HealthTrackerType.ACCELEROMETER_CONTINUOUS, AccListener(dao, listenerProxy))
                trackerManager.startContinuous(
                    HealthTrackerType.HEART_RATE_CONTINUOUS, HrListener(dao, listenerProxy))
                trackerManager.startContinuous(
                    HealthTrackerType.PPG_CONTINUOUS, PpgListener(dao, listenerProxy))
                trackerManager.startContinuous(
                    HealthTrackerType.SKIN_TEMPERATURE_CONTINUOUS, SkinTempListener(dao, listenerProxy))
                scheduler.start()
            }

            override fun onConnectionFailed(e: HealthTrackerException) {
                Log.e(TAG, "Conexión fallida (errorCode=${e.errorCode})", e)
                serviceScope.launch {
                    Toast.makeText(
                        this@ForegroundService,
                        getString(R.string.msg_connection_failed),
                        Toast.LENGTH_LONG
                    ).show()
                }
                scheduleReconnect()
            }

            override fun onConnectionEnded() {
                Log.w(TAG, "HealthTrackingService desconectado — programando reconexión")
                trackerManager.markDisconnected()
                scheduler.stop()
                scheduleReconnect()
            }
        }

        trackerManager = TrackerManager(this, connListener)
        scheduler      = OnDemandScheduler(trackerManager, dao, listenerProxy)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        createNotificationChannel()
        val notif: Notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(getString(R.string.notif_title))
            .setContentText(getString(R.string.notif_text))
            .setSmallIcon(R.drawable.ic_launcher_foreground)
            .build()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            startForeground(NOTIF_ID, notif,
                android.content.pm.ServiceInfo.FOREGROUND_SERVICE_TYPE_HEALTH)
        } else {
            startForeground(NOTIF_ID, notif)
        }

        wakeLock?.takeIf { !it.isHeld }?.acquire(WAKELOCK_TIMEOUT_MS)

        running = true
        return START_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()
        running = false
        serviceJob.cancel()
        wakeLock?.takeIf { it.isHeld }?.release()
        scheduler.destroy()
        trackerManager.stopAllContinuous()
        trackerManager.shutdown()
    }

    // ── Binder ────────────────────────────────────────────────────────────────
    inner class LocalBinder : Binder() {
        fun getService(): ForegroundService = this@ForegroundService
    }

    private val binder = LocalBinder()
    override fun onBind(intent: Intent): IBinder = binder

    // ── API pública ───────────────────────────────────────────────────────────
    fun exportCsv(filename: String) {
        serviceScope.launch(Dispatchers.IO) {
            try {
                CsvExporter.export(dao, this@ForegroundService, filename)
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@ForegroundService,
                        getString(R.string.msg_csv_saved, filename), Toast.LENGTH_SHORT).show()
                }
            } catch (e: IOException) {
                Log.e(TAG, "Error exportando CSV", e)
                withContext(Dispatchers.Main) {
                    Toast.makeText(this@ForegroundService,
                        getString(R.string.msg_csv_error), Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    // ── Reconexión con backoff exponencial ────────────────────────────────────
    private fun scheduleReconnect() {
        if (!running) return
        val delayMs = (BASE_RECONNECT_MS shl reconnectAttempt.coerceAtMost(MAX_SHIFT))
            .coerceAtMost(MAX_RECONNECT_MS)
        reconnectAttempt++
        Log.i(TAG, "Reconectando en ${delayMs}ms (intento $reconnectAttempt)")
        serviceScope.launch {
            delay(delayMs)
            if (running) trackerManager.reconnect()
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────
    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val channel = NotificationChannel(
            CHANNEL_ID,
            getString(R.string.notif_channel_name),
            NotificationManager.IMPORTANCE_LOW
        ).apply { description = getString(R.string.notif_channel_desc) }
        getSystemService(NotificationManager::class.java)?.createNotificationChannel(channel)
    }

    companion object {
        private const val TAG = "ForegroundService"
        private const val CHANNEL_ID = "sensors_channel"
        private const val NOTIF_ID = 1
        private const val WAKELOCK_TIMEOUT_MS = 12L * 60 * 60 * 1000

        // Backoff: 2s, 4s, 8s, 16s, 32s, 60s (tope)
        private const val BASE_RECONNECT_MS = 2_000L
        private const val MAX_RECONNECT_MS  = 60_000L
        private const val MAX_SHIFT         = 5       // 2^5 * 2s = 64s > tope

        @Volatile private var running = false
        fun isRunning() = running
    }
}
